src/pipecat/services/tavus.py

```diff
@@ -0,0 +1,140 @@
+#
+# Copyright (c) 2024, Daily
+#
+# SPDX-License-Identifier: BSD 2-Clause License
+#
+
+
+"""This module implements Tavus as a sink transport layer"""
+
+import aiohttp
+import base64
+
+from pipecat.frames.frames import (
+    Frame,
+    TTSAudioRawFrame,
+    TransportMessageUrgentFrame,
+    TTSStartedFrame,
+    TTSStoppedFrame,
+    StartInterruptionFrame,
+)
+from pipecat.processors.frame_processor import FrameDirection, FrameProcessor
+
+from loguru import logger
+
+
+class TavusVideoService(FrameProcessor):
+    """Class to send base64 encoded audio to Tavus"""
+
+    def __init__(
+        self,
+        *,
+        conversation_id: str,
+        **kwargs,
+    ) -> None:
+        super().__init__(**kwargs)
+        self._conversation_id = conversation_id
+
+    def can_generate_metrics(self) -> bool:
+        return True
+
+    @classmethod
+    async def get_persona_name(
+        cls,
+        *,
+        session: aiohttp.ClientSession,
+        api_key: str,
+        persona_id: str,
+    ) -> str:
+        url = f"https://tavusapi.com/v2/personas/{persona_id}"
+        headers = {"Content-Type": "application/json", "x-api-key": api_key}
+        async with session.get(url, headers=headers) as r:
+            r.raise_for_status()
+            response_json = await r.json()
+
+        logger.debug(f"TavusVideoService persona grabbed {response_json}")
+        return response_json["persona_name"]
+
+    @classmethod
+    async def end_conversation(
+        cls,
+        *,
+        session: aiohttp.ClientSession,
+        api_key: str,
+        conversation_id: str,
+    ):
+        url = f"https://tavusapi.com/v2/conversations/{conversation_id}/end"
+        headers = {"Content-Type": "application/json", "x-api-key": api_key}
+        async with session.post(url, headers=headers) as r:
+            r.raise_for_status()
+
+    @classmethod
+    async def initiate_conversation(
+        cls,
+        *,
+        session: aiohttp.ClientSession,
+        api_key: str,
+        replica_id: str,
+        persona_id: str = "pipecat0",
+    ) -> tuple[str, str]:
```

Comment:
Is there any reason for `initiate_conversation` and `end_conversation`? Maybe we could just pass `api_key`, `persona_id`, `replica_id` and `aiohttp_session` to the constructor and do all this internally? We then could use `StartFrame` to initiate conversation and also `FrameProcessor.cleanup()` to end the conversation? Something like this:

```
    def __init__(
        self,
        *,
        persona_id: str,
        replica_id: str,
        **kwargs,
    ) -> None:
```

and then

```
    async def process_frame(self, frame: Frame, direction: FrameDirection):
        await super().process_frame(frame, direction)
        if isinstance(frame, StartFrame):
            await self.push_frame(frame, direction)
            await self._start(frame)
        ....
        
    async def _start(self, frame: StartFrame):
        self._conversation_id = await self._initiate_conversation()

    async def cleanup(self):
        self._end_conversation()        
```

This way the user doesn't need to call `initiate_conversation` and `end_conversation` it's all done automatically.

`get_persona_name` is fine.

Reply:
See other example about `get_persona_name`. If we pass the api key, etc to the constructor it might be just simpler to have a member function so we don't have to write that much code (see other example)

Reply:
The issue with this is we need `room_url` beforehand so we can pass it to the `DailyTransport`

Reply:
I put `tavus.end_conversation()` inside a `if isinstance(frame, EndFrame):` but left `_start()` outside (i.e., `room_url = await tavus.initialize()` is called in `examples/foundational/21-tavus-layer.py`)